### PR TITLE
fix(array): make docarrays sync when explicitly synced and when exiting context manager only

### PR DESCRIPTION
**Pull request by Jackmin801**
Goals:

- [x] Fix issue mentioned in https://github.com/jina-ai/docarray/issues/529
- [x] Address changes proposed in https://github.com/jina-ai/docarray/pull/604
- [x] Add fix to base so it propagates to other backends
- [x] Make sure other backends aren't broken
- [x] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

File: docarray/array/storage/annlite/getsetdel.py [hunk 1]
```diff
@@ -42,15 +42,15 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
     def _del_docs_by_ids(self, ids):
         self._annlite.delete(ids)
 
+    def sync(self) -> None:
```
Comment (JoanFM) [inline docarray/array/storage/annlite/getsetdel.py:45]: if u do not need to override, no need to add it here

File: docarray/array/storage/sqlite/seqlike.py [hunk 2]
```diff
@@ -48,8 +48,8 @@ def _append(self, doc: 'Document', commit: bool = True, **kwargs) -> None:
         if commit:
             self._commit()
 
-    def __del__(self) -> None:
-        super().__del__()
+    def sync(self) -> None:
+        super().sync()
```
Comment (JoanFM) [inline docarray/array/storage/sqlite/seqlike.py:52]: also here

File: tests/unit/array/test_advance_indexing.py [hunk 3]
```diff
@@ -680,7 +680,7 @@ def test_edge_case_two_strings(storage, config_gen, start_storage):
     [
         ('sqlite', None),
         ('weaviate', WeaviateConfig(n_dim=123)),
-        ('annlite', AnnliteConfig(n_dim=123)),
+        # ('annlite', AnnliteConfig(n_dim=123)),
```
Comment (JoanFM) [inline tests/unit/array/test_advance_indexing.py:683]: why?

File: tests/unit/array/test_advance_indexing.py [hunk 4]
```diff
@@ -689,26 +689,27 @@ def test_edge_case_two_strings(storage, config_gen, start_storage):
...
     da = DocumentArray(storage=storage, config=config)
 
-    da.extend(
-        [
-            Document(id='0'),
-            Document(id='2'),
-            Document(id='4'),
-        ]
-    )
+    da.extend([Document(id=i) for i in '024'])
```
Comment (JoanFM) [inline tests/unit/array/test_advance_indexing.py:692]: I prefer the old way, more explicit

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/array/test_advance_indexing.py (hunk 3 above)
Comment (Jackmin801) [reply tests/unit/array/test_advance_indexing.py:683]: Looking into it. Something with it not closing properly

Comment (Jackmin801) [reply tests/unit/array/test_advance_indexing.py:683]: Then it conflicts its own lock in the second reference

File: docarray/array/storage/annlite/getsetdel.py (hunk 1 above)
Comment (Jackmin801) [reply docarray/array/storage/annlite/getsetdel.py:45]: Oh I hadnt realized this was all that was left here 😅 good catch

File: docarray/array/storage/sqlite/seqlike.py (hunk 2 above)
Comment (Jackmin801) [reply docarray/array/storage/sqlite/seqlike.py:52]: Oh there is still the if statement after

File: docarray/array/storage/weaviate/seqlike.py [hunk 5]
```diff
@@ -54,9 +54,9 @@ def __contains__(self, x: Union[str, 'Document']):
         else:
             return False
 
-    def __del__(self):
+    def sync(self):
         """Delete this :class:`DocumentArrayWeaviate` object"""
-        super().__del__()
+        super().sync()
```
Comment (Jackmin801) [inline docarray/array/storage/weaviate/seqlike.py:59]: Why does it not think these are covered by tests? Isn't tests/unit/array/test_advance_indexing.py::test_offset2ids_persistence explicitly calling sync on this DocumentStore?

Comment (Jackmin801) [reply docarray/array/storage/weaviate/seqlike.py:59]: Oh its not being overriden... thanks codecov <3

**Review by JoanFM — CHANGES_REQUESTED** [review]
Please, remember to clarify in Documentation

**Review by JoanFM — COMMENTED** [review]
Add Documentation about `sync` intergace

**Review by JoanFM — APPROVED** [review]

File: docs/advanced/document-store/index.md [hunk 6]
```diff
@@ -145,6 +145,14 @@ da = DocumentArray(
 
 Using dataclass gives you better type-checking in IDE but requires an extra import; using dict is more flexible but can be error-prone. You can choose the style that fits best to your context.
 
+```{admonition} Creating DocumentArrays without specifying index
+:class: warning
+When you specify an index (table name for SQL stores) in the config, the index will be used to persist the DocumentArray in the document store.
+If you create a DocumentArray but do not specify an index, a randomized placeholder index will be created to persist the data.
+
+Creating DocumentArrays without index is useful during prototyping but should not be used in a production setting as randomized placeholder data will be persisted in the document store unnecessarily.
```
Comment (NicholasDunham) [inline docs/advanced/document-store/index.md:153]:
```suggestion
Creating DocumentArrays without indexes is useful during prototyping but should not be used in a production setting as randomized placeholder data will be persisted in the document store unnecessarily.
```

**Review by NicholasDunham — CHANGES_REQUESTED** [review]
One minor change; otherwise looks good

**Review by JoanFM — APPROVED** [review]

**Pull request merged**